lib/kcerts/ssh.go

```diff
@@ -0,0 +1,199 @@
+
+var (
+	sockR = regexp.MustCompile("(?m)SSH_AUTH_SOCK=([^;\\n]*)")
+	pidR  = regexp.MustCompile("(?m)SSH_AGENT_PID=([0-9]*)")
+)
+
+// FindSSHDir will find the users ssh directory based on $HOME. If $HOME/.ssh does not exist
+// it will attempt to create it.
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	defer knownHostsFile.Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type SSHAgent struct {
+	PID    int    `json:"pid"`
+	Socket string `json:"sock"`
+	// failed is marked in WriteToCache
+	failed bool
+}
+
+func (a SSHAgent) Valid() bool {
+	conn, err := net.Dial("unix", a.Socket)
+	defer conn.Close()
+	return err == nil
+}
+
+// CloseIfNotCached will kill the agent if it failed to write to cache. Otherwise is a NoOp
+func (a *SSHAgent) CloseIfNotCached() error {
+	if a.failed {
+		cmd := exec.Command("kill", "-HUP", strconv.Itoa(a.PID))
+		cmd.Stdout = os.Stdout
+		return cmd.Run()
+	}
+	return nil
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, logger *klog.Logger) (*SSHAgent, error) {
+	// TODO(adam): sniff out dead agents and sneak pids
+	agent := FindSSHAgentFromEnv()
+	if agent != nil && agent.Valid() {
+		return agent, nil
+	}
+	agent, err := FetchSSHAgentFromCache(store)
+	if err != nil {
+		logger.Warn(err.Error())
+		if errors.Is(SSHAgentCacheInvalid, err) {
```

Comment:
I would remove this block, from line 111 to 115. It's none of the caller's business to cleanup after FetchSSHAgentFromCache own mess, and ... due to how the cache is structured, sure enough this file will be overwritten the first time we succeed writing in the cache, don't bother removing the entry.

(fun fact: if one was using nfs or bind mounts, the cached file may be valid in one terminal, invalid in another - I recommend being lazy here)